x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,162 @@
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Random;
+import java.util.Set;
+import java.util.function.Supplier;
+
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+public class ResultsPersisterService {
+    private static final Logger LOGGER = LogManager.getLogger(ResultsPersisterService.class);
+
+    public static final Setting<Integer> PERSIST_RESULTS_MAX_RETRIES = Setting.intSetting(
+        "xpack.ml.persist_results_max_retries",
+        15,
+        0,
+        50,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope);
+    private static final int MAX_RETRY_SLEEP_MILLIS = 900_000; // 15 min
+    private static final int MIN_RETRY_SLEEP_MILLIS = 50;
+    // Having an exponent higher than this causes integer overflow
+    private static final int MAX_RETRY_EXPONENT = 29;
+
+    private final Random random = Randomness.get();
+    private final Client client;
+    private volatile int maxFailureRetries;
+
+    public ResultsPersisterService(Client client, ClusterService clusterService, Settings settings) {
+        this.client = client;
+        this.maxFailureRetries = PERSIST_RESULTS_MAX_RETRIES.get(settings);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(PERSIST_RESULTS_MAX_RETRIES, this::setMaxFailureRetries);
+    }
+
+    void setMaxFailureRetries(int value) {
+        this.maxFailureRetries = value;
+    }
+
+    public BulkResponse indexWithRetry(String jobId,
+                                       String indexName,
+                                       ToXContent object,
+                                       ToXContent.Params params,
+                                       WriteRequest.RefreshPolicy refreshPolicy,
+                                       String id,
+                                       Supplier<Boolean> shouldRetry) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        try (XContentBuilder content = object.toXContent(XContentFactory.jsonBuilder(), params)) {
+            bulkRequest.add(new IndexRequest(indexName).id(id).source(content));
+        }
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry);
+    }
+
+    public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest, String jobId, Supplier<Boolean> shouldRetry) {
+        int currentMin = MIN_RETRY_SLEEP_MILLIS;
+        int currentMax = MIN_RETRY_SLEEP_MILLIS;
+        int currentAttempt = 0;
+        while(currentAttempt <= maxFailureRetries) {
+            BulkResponse bulkResponse = bulkIndex(bulkRequest);
+            if (bulkResponse.hasFailures() == false) {
+                return bulkResponse;
+            }
+            if (shouldRetry.get() == false) {
+                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+            }
+            if (currentAttempt > maxFailureRetries) {
```

Comment:
Do I understand correctly that this condition will evaluate to `true` only if the cluster-level setting `maxFailureRetries` is changed to be lower than before?
If so, I think the warning message should mention that. If not, then should incrementing `currentAttempt` happen before this line rather than in line 98?